blockchain/utxocache.go

```diff
@@ -235,6 +235,125 @@ type utxoBatcher interface {
 	getEntries(outpoints []wire.OutPoint) ([]*UtxoEntry, error)
 }
 
+// mapSlice is a slice of maps for Utxo entries.  The slice of maps are needed to
+// guarantee that the map will only take up N amount of bytes.  As of v1.20, the
+// go runtime will allocate 2^N + few extra buckets, meaning that for large N, we'll
+// allocate a lot of extra memory if the amount of entries goes over the previously
+// allocated buckets.  A slice of maps allows us to have a better control of how much
+// total memory gets allocated by all the maps.
+type mapSlice struct {
+	maps []map[wire.OutPoint]*UtxoEntry
+
+	// maxEntries is the maximum amount of elemnts that the map is allocated for.
+	maxEntries []int
+
+	// maxTotalMemoryUsage is the maximum memory usage in bytes that the state
+	// should contain in normal circumstances.
+	maxTotalMemoryUsage uint64
+}
+
+// length returns the length of all the maps in the map slice added together.
+func (ms *mapSlice) length() int {
+	var l int
+	for _, m := range ms.maps {
+		l += len(m)
+	}
+
+	return l
+}
+
+// size returns the size of all the maps in the map slice added together.
+func (ms *mapSlice) size() int {
+	var size int
+	for _, num := range ms.maxEntries {
+		size += CalculateRoughMapSize(num, bucketSize)
+	}
+
+	return size
+}
+
+// get looks for the outpoint in all the maps in the map slice and returns
+// the entry.  nil and false is returned if the outpoint is not found.
+func (ms *mapSlice) get(op wire.OutPoint) (*UtxoEntry, bool) {
+	var entry *UtxoEntry
+	var found bool
+
+	for _, m := range ms.maps {
```

Comment:
Kinda feels like we're just reinventing hashmap bucket operations (multi-look up, overflow, etc) but on a "user space" level vs letting the runtime just handle it. 

I wonder if the maps themselves can be ordered based on recency, so assumption that check the lowest level first as that's where the newest UTXOs are found. This would require actually doing eviction though.

Reply:
> Kinda feels like we're just reinventing hashmap bucket operations

Yeah... true but if we let the runtime handle it it'll allocate something that's bigger than what the user specified with `--utxocachemaxsize`.

> I wonder if the maps themselves can be ordered based on recency, so assumption that check the lowest level first as that's where the newest UTXOs are found. This would require actually doing eviction though.

At the moment the maps are ordered from largest to smallest. And since the smaller maps are allocated later, the general order is oldest to youngest. I guess you could look up the smaller maps first. Maybe that'll have a positive impact?